File: lib/supabase.py
"""Client REST minimale per Supabase (PostgREST), stdlib puro — nessun `requests`,
nessun `supabase-py`. Solo `urllib`, come già fa il motore outreach.

Le tabelle `sheis_*` sono definite in ~/alkemia-sheis-backend/migrations/0001 e 0002,
ma **non esistono ancora nel progetto reale** (verificato 2026-08-03: manca il
Personal Access Token per il DDL via Management API — la sola service key legge/scrive
RIGHE via PostgREST ma non crea tabelle). Ogni worker DEVE partire lo stesso e
DICHIARARE questo stato invece di andare in crash con un errore tecnico: è il compito
di `schema_pronto()` qui sotto.
"""
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from pathlib import Path

TIMEOUT = 30

# Tabelle che i worker di questo repo si aspettano di trovare (0001 + 0002).
TABELLE_ATTESE = [
    "sheis_utenti", "sheis_contenuti", "sheis_varianti", "sheis_pubblicazioni",
    "sheis_candidati", "sheis_campagne", "sheis_articoli", "sheis_report",
    "sheis_approvazioni_log", "sheis_distributori", "sheis_catalogo",
]


def _candidati_env() -> list[Path]:
    """.env di QUESTO repo prima di tutto; poi (stesso cliente, stesso progetto
    Supabase) il .env di alkemia-sheis-backend, che possiede lo schema. Non si
    risale a nient'altro: è lo stesso pattern-lezione di canali.py — solo file
    che appartengono allo STESSO cliente/progetto, mai a un altro.
    """
    qui = Path(__file__).resolve().parent.parent
    return [
        Path(os.environ.get("SHEIS_ENV_FILE", "")) if os.environ.get("SHEIS_ENV_FILE") else None,
        qui / ".env",
        Path(os.environ.get("SHEIS_BACKEND_ENV", str(Path.home() / "alkemia-sheis-backend" / ".env"))),
    ]


def _carica_env() -> None:
    if os.environ.get("SUPABASE_URL") and os.environ.get("SUPABASE_SECRET_KEY"):
        return
    for p in _candidati_env():
        if not p or not p.is_file():
            continue
        for riga in p.read_text(encoding="utf-8").splitlines():
            riga = riga.strip()
            if not riga or riga.startswith("#") or "=" not in riga:
                continue
            k, v = riga.split("=", 1)
            k = k.strip()
            if k in ("SUPABASE_URL", "SUPABASE_SECRET_KEY", "SUPABASE_PROJECT_REF") and k not in os.environ:
                os.environ[k] = v.strip().strip('"').strip("'")
        if os.environ.get("SUPABASE_URL") and os.environ.get("SUPABASE_SECRET_KEY"):
            return


@dataclass
class Esito:
    ok: bool
    dati: list | dict = field(default_factory=list)
    errore: str = ""
    schema_mancante: bool = False


class SupabaseClient:
    def __init__(self) -> None:
        _carica_env()
        self.url = (os.environ.get("SUPABASE_URL") or "").rstrip("/")
        self.chiave = os.environ.get("SUPABASE_SECRET_KEY") or ""
        self.credenziali_presenti = bool(self.url and self.chiave)

    # ------------------------------------------------------------ richiesta grezza
    def _req(self, method: str, path: str, body: dict | list | None = None,
              extra_headers: dict | None = None) -> Esito:
        if not self.credenziali_presenti:
            return Esito(ok=False, errore="SUPABASE_URL/SUPABASE_SECRET_KEY assenti — vedi .env.example")

        headers = {
            "apikey": self.chiave,
            "Authorization": f"Bearer {self.chiave}",
            "Content-Type": "application/json",
        }
        headers.update(extra_headers or {})
        data = json.dumps(body).encode("utf-8") if body is not None else None
        req = urllib.request.Request(f"{self.url}{path}", data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
                grezzo = r.read().decode("utf-8")
                return Esito(ok=True, dati=json.loads(grezzo) if grezzo else [])
        except urllib.error.HTTPError as e:
            corpo = e.read().decode("utf-8", errors="replace")
            schema_mancante = e.code in (400, 404) and (
                "PGRST205" in corpo or "does not exist" in corpo or "relation" in corpo.lower()
            )
            return Esito(ok=False, errore=f"HTTP {e.code}: {corpo[:300]}", schema_mancante=schema_mancante)
        except urllib.error.URLError as e:
            return Esito(ok=False, errore=f"rete: {e}")
        except Exception as e:  # json malformato, timeout residuo, ecc.
            return Esito(ok=False, errore=f"{type(e).__name__}: {e}")

    # ------------------------------------------------------------------- CRUD
    def select(self, tabella: str, query: str = "select=*", limit: int | None = None) -> Esito:
        qs = query
        if limit is not None:
            qs += f"&limit={limit}"
        return self._req("GET", f"/rest/v1/{tabella}?{qs}")

    def insert(self, tabella: str, righe: list[dict] | dict) -> Esito:
        return self._req("POST", f"/rest/v1/{tabella}", body=righe,
                          extra_headers={"Prefer": "return=representation"})

    def upsert(self, tabella: str, righe: list[dict] | dict, conflitto: str) -> Esito:
        return self._req("POST", f"/rest/v1/{tabella}?on_conflict={conflitto}", body=righe,
                          extra_headers={"Prefer": "return=representation,resolution=merge-duplicates"})

    def update(self, tabella: str, filtro: str, patch: dict) -> Esito:
        return self._req("PATCH", f"/rest/v1/{tabella}?{filtro}", body=patch,
                          extra_headers={"Prefer": "return=representation"})

    # ------------------------------------------------------ stato dello schema
    def tabella_esiste(self, tabella: str) -> bool:
        esito = self.select(tabella, query="select=id", limit=0)
        return esito.ok

    def schema_pronto(self, tabelle: list[str] | None = None) -> tuple[bool, str]:
        """(pronto, messaggio in italiano). Non solleva mai: un DB non ancora
        inizializzato è uno STATO da dichiarare, non un errore tecnico da esporre.
        """
        if not self.credenziali_presenti:
            return False, (
                "database non ancora inizializzato: mancano SUPABASE_URL/SUPABASE_SECRET_KEY "
                "(vedi .env.example — copiali da ~/alkemia-sheis-backend/.env)"
            )
        attese = tabelle or TABELLE_ATTESE
        mancanti = [t for t in attese if not self.tabella_esiste(t)]
        if mancanti:
            return False, (
                "database non ancora inizializzato: mancano le tabelle "
                f"{', '.join(mancanti)} (le migrazioni 0001/0002 non sono state applicate — "
                "serve un Personal Access Token Supabase per il DDL, vedi "
                "~/alkemia-sheis-backend/applica_migrazioni.py)"
            )
        return True, f"schema pronto — {len(attese)} tabelle verificate"

File: studio/src/lib/supabase.ts
import "server-only";

/**
 * Client Supabase minimale, server-only, via REST + chiave di servizio.
 * Nessun SDK pesante: la superficie che serve è piccola (select/insert/update)
 * e un client scritto a mano rende ESPLICITO il punto in cui rilevare lo
 * "schema non inizializzato" — la condizione reale del progetto oggi
 * (0/15 tabelle sheis_*, manca il Personal Access Token per il DDL).
 *
 * Non importare MAI questo file da un componente client: usa solo da route
 * API e Server Component. La chiave di servizio non deve mai arrivare al browser.
 */

const SUPABASE_URL = process.env.SUPABASE_URL ?? "";
const SERVICE_KEY = process.env.SUPABASE_SECRET_KEY ?? "";

/** Lo schema sheis_* non esiste ancora in questo progetto Supabase. */
export class SchemaNotInitializedError extends Error {
  table: string;
  constructor(table: string) {
    super(
      `La tabella "${table}" non esiste ancora: lo schema del database non è stato inizializzato.`,
    );
    this.name = "SchemaNotInitializedError";
    this.table = table;
  }
}

/** Manca la configurazione minima (URL o chiave di servizio) in .env.local. */
export class SupabaseConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SupabaseConfigError";
  }
}

/** Supabase ha risposto con un errore che non è "tabella mancante". */
export class SupabaseRequestError extends Error {
  status: number;
  constructor(status: number, message: string) {
    super(message);
    this.name = "SupabaseRequestError";
    this.status = status;
  }
}

function assertConfigured(): void {
  if (!SUPABASE_URL || !SERVICE_KEY) {
    throw new SupabaseConfigError(
      "Configurazione Supabase mancante: imposta SUPABASE_URL e SUPABASE_SECRET_KEY in .env.local.",
    );
  }
}

type Metodo = "GET" | "POST" | "PATCH" | "DELETE";

export type RestOptions = {
  method?: Metodo;
  /** Querystring PostgREST già composta, es. "select=*&stato=eq.approvato&order=created_at.desc". */
  query?: string;
  body?: unknown;
  /** Header Prefer, es. "return=representation" per riavere le righe scritte. */
  prefer?: string;
};

/**
 * Chiamata REST diretta a PostgREST. Lancia SchemaNotInitializedError quando
 * Supabase risponde col codice PGRST205 (tabella non trovata nello schema
 * cache) — è la firma esatta misurata su questo progetto il 2026-08-03.
 */
export async function sbFetch<T>(table: string, opts: RestOptions = {}): Promise<T> {
  assertConfigured();
  const { method = "GET", query = "", body, prefer } = opts;
  const url = `${SUPABASE_URL}/rest/v1/${table}${query ? `?${query}` : ""}`;

  const headers: Record<string, string> = {
    apikey: SERVICE_KEY,
    authorization: `Bearer ${SERVICE_KEY}`,
    "content-type": "application/json",
  };
  if (prefer) headers.prefer = prefer;

  let r: Response;
  try {
    r = await fetch(url, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
      cache: "no-store",
    });
  } catch {
    throw new SupabaseRequestError(0, "Errore di rete verso Supabase. Riprova tra poco.");
  }

  if (!r.ok) {
    const text = await r.text().catch(() => "");
    let code: string | undefined;
    try {
      code = (JSON.parse(text) as { code?: string }).code;
    } catch {
      /* corpo non JSON: ignora, resta il testo grezzo nel messaggio d'errore */
    }
    if (r.status === 404 && code === "PGRST205") {
      throw new SchemaNotInitializedError(table);
    }
    throw new SupabaseRequestError(
      r.status,
      `Supabase ha risposto ${r.status} su "${table}": ${text.slice(0, 300) || "(nessun dettaglio)"}`,
    );
  }

  if (r.status === 204) return undefined as T;
  const text = await r.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
}

export type StatoSchema = { ok: boolean; motivo?: string };

/**
 * Verifica se lo schema Studio risponde, senza mai lanciare. È la funzione da
 * chiamare a ogni pagina/route che tocca il database, per decidere se
 * mostrare il prodotto o dichiarare "database non ancora inizializzato".
 */
export async function schemaInizializzato(): Promise<StatoSchema> {
  if (!SUPABASE_URL || !SERVICE_KEY) {
    return {
      ok: false,
      motivo: "Configurazione Supabase mancante: imposta SUPABASE_URL e SUPABASE_SECRET_KEY in .env.local.",
    };
  }
  try {
    await sbFetch("sheis_utenti", { query: "select=id&limit=1" });
    return { ok: true };
  } catch (e) {
    if (e instanceof SchemaNotInitializedError) {
      return {
        ok: false,
        motivo:
          "Il database non è ancora stato inizializzato: le tabelle sheis_* non esistono su Supabase. " +
          "Serve applicare le migrazioni in ~/alkemia-sheis-backend/migrations/ (richiede un Personal Access Token Supabase, oggi mancante).",
      };
    }
    if (e instanceof SupabaseConfigError) return { ok: false, motivo: e.message };
    return {
      ok: false,
      motivo: e instanceof Error ? e.message : "Errore sconosciuto nel controllo del database.",
    };
  }
}
